### PR TITLE
feat(azure-handler): implement ProgramIngest HTTP trigger endpoint

### PR DESCRIPTION
**Pull request by Alan-Jowett**
## Summary

Implements the `ProgramIngest` Azure Function endpoint (WEB-0300) that accepts ELF binaries via JSON POST, verifies them with Prevail, encodes deterministic CBOR program images, and stores them in Azure Tables for delivery via `DESIRED_STATE` messages.

### `sonde-azure-handler` (lib.rs)
- `ProgramImageRow` struct and `HandlerStore::store_program_image()` trait method
- `AzureHandler::handle_program_ingest()` — JSON body parsing, Prevail verification, CBOR encoding, storage
- `IngestResponse`/`IngestError` types with HTTP status mapping (400/413/422/500)
- `extract_http_trigger_body()` and `format_ingest_response()` for Azure Functions envelope handling
- Input validation: base64, empty check, 1 MB ELF limit, `abi_version` type/range, `verification_profile` enum
- Unit tests covering T-WEB-0301 through T-WEB-0306, T-WEB-0307a (empty ELF), and T-WEB-0308. T-WEB-0304 (integration) and T-WEB-0307b (multi-program ELF) remain planned.

### `sonde-azure-handler` (main.rs)
- Dedicated `/ProgramIngest` route

### Specification updates
- `gateway-companion-api.md`: Documented key 5 (`assigned_program_image`) in DESIRED_STATE
- `web-ui-design.md`: Updated sections 6.1 and 9.3 with implementation details
- `web-ui-validation.md`: Updated test statuses

## Test results

66 tests pass, 0 failures. Clippy clean across workspace.

### CODE REVIEW

File: crates/sonde-azure-handler/function-app/ProgramIngest/function.json [hunk 1]
```diff
@@ -11,7 +11,7 @@
     {
       "type": "http",
       "direction": "out",
-      "name": "$return"
+      "name": "res"
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply crates/sonde-azure-handler/function-app/ProgramIngest/function.json:14]: Fixed — added `ProgramIngest/function.json` to the Dockerfile package build.

File: crates/sonde-azure-handler/src/lib.rs [hunk 2]
```diff
@@ -322,6 +339,281 @@ where
...
+                Some(val)
+            }
+        };
+
+        let lib = ProgramLibrary::new();
+        let mut record = lib.ingest_elf(&elf_bytes, profile).map_err(|e| {
+            IngestError::unprocessable(format!(
+                "program verification/ingestion failed: {e}"
+            ))
+        })?;
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply crates/sonde-azure-handler/src/lib.rs:446]: Fixed -- ProgramError::Internal now maps to HTTP 500; other variants remain 422.

File: docs/web-ui-design.md [hunk 3]
```diff
@@ -148,11 +198,19 @@ After deployment, grant users the `Storage Table Data Contributor` role on the s
...
+- `ProgramIngest/function.json` defines the HTTP trigger (`authLevel: function`,
+  route `programs/ingest`).
+- `main.rs` routes `/ProgramIngest` to a dedicated handler that parses the
+  Azure Functions HTTP trigger envelope and delegates to
+  `AzureHandler::handle_program_ingest()`. The catch-all `/{*path}` route
+  continues to handle queue-triggered connector messages.
+- `lib.rs` implements `handle_program_ingest()` which reuses
+  `ProgramLibrary::ingest_elf()` from `sonde-gateway` for Prevail verification,
+  CBOR encoding, and SHA-256 hashing.
+- The HTTP trigger response uses the Azure Functions `$return` output binding
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/web-ui-design.md]: Fixed -- updated doc to reference the res output binding name.

File: docs/web-ui-design.md [hunk 4]
```diff
@@ -69,15 +69,63 @@ deploy/web-ui/
...
+  2. Base64-decode the `elf` field; reject if missing, empty, or invalid base64
+  3. Reject requests where decoded ELF exceeds 1 MB (defense-in-depth before verification)
+  4. Parse `verification_profile`; default to `Resident` if absent
+  5. Call `ProgramLibrary::ingest_elf(elf_bytes, profile)` — this verifies with Prevail, extracts bytecode + maps, encodes deterministic CBOR, and computes SHA-256
+  6. Normalize `source_filename` via `normalize_display_filename()`
+  7. Store in `programs` Azure Table via `HandlerStore::store_program_image()`
+  8. Return JSON: `{"program_hash": "hex", "size": N, "abi_version": N, "source_filename": "name"}`
+  9. On failure: return JSON error with diagnostics
+- The uploaded ELF is verified and transformed; the persisted and delivered artifact is the deterministic CBOR program image, not the raw ELF
+- `programs` table schema: `PartitionKey="program"`, `RowKey=hex(program_hash)`, `source_filename`, `abi_version` (`Edm.Int32`), `cbor_image` (base64-encoded CBOR program image), `size_bytes` (`Edm.Int32`, CBOR image byte length), `verification_profile`, `created_at` (`Edm.DateTime`)
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/web-ui-design.md]: Fixed -- changed spec to created_at (ISO 8601 UTC string) to match the implementation.

File: crates/sonde-azure-handler/src/lib.rs [hunk 5]
```diff
@@ -322,6 +339,281 @@ where
...
+                "ELF size {} bytes exceeds limit of {} bytes",
+                elf_bytes.len(),
+                MAX_ELF_UPLOAD_SIZE
+            )));
+        }
+
+        let profile_str = body
+            .get("verification_profile")
+            .and_then(|v| v.as_str())
+            .unwrap_or("resident");
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply crates/sonde-azure-handler/src/lib.rs]: Fixed -- non-string verification_profile values now return 400. Only absent/null defaults to resident. Added test.

File: crates/sonde-azure-handler/src/lib.rs [hunk 6]
```diff
@@ -1152,6 +1480,14 @@ mod tests {
                 .get(&hex::encode(program_hash))
                 .cloned())
         }
+
+        async fn store_program_image(&self, row: &ProgramImageRow) -> Result<(), HandlerError> {
+            self.program_images
+                .lock()
+                .await
+                .insert(hex::encode(&row.program_hash), row.cbor_image.clone());
+            Ok(())
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply crates/sonde-azure-handler/src/lib.rs:1497]: Fixed -- MemoryStore now captures the full ProgramImageRow. Test asserts all metadata fields (source_filename, abi_version, size_bytes, verification_profile, created_at, cbor_image).

File: docs/web-ui-validation.md [hunk 7]
```diff
@@ -22,7 +22,7 @@
 | T-WEB-0203 | WEB-0203 | RowKey uses reverse-timestamp format | Unit (JS) | Planned |
 | T-WEB-0204 | WEB-0204 | PartitionKey = `n:{SHA-256(node_id)}` | Unit (JS) | Planned |
 | T-WEB-0205 | WEB-0205 | `timestamp_ms` stored as `Edm.Int64` | Integration | Planned |
-| T-WEB-0301 | WEB-0301 | `ProgramIngest` accepts ELF + metadata via HTTP POST | Unit (Rust) | Planned |
+| T-WEB-0301 | WEB-0301 | `ProgramIngest` accepts ELF + metadata via JSON POST | Unit (Rust) | Planned |
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/web-ui-validation.md]: Fixed -- updated T-WEB-0301 through T-WEB-0308 statuses from Planned to Pass. Changed T-WEB-0304 method from Integration to Unit (Rust).

File: crates/sonde-azure-handler/src/lib.rs [hunk 8]
```diff
@@ -322,6 +339,281 @@ where
...
+        }
+
+        let profile_str = body
+            .get("verification_profile")
+            .and_then(|v| v.as_str())
+            .unwrap_or("resident");
+        let profile = match profile_str {
+            "resident" => VerificationProfile::Resident,
+            "ephemeral" => VerificationProfile::Ephemeral,
+            other => {
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply crates/sonde-azure-handler/src/lib.rs:399]: Fixed -- added program_ingest_ephemeral_profile test that exercises the ephemeral branch and asserts the stored profile.

File: crates/sonde-azure-handler/src/lib.rs [hunk 9]
```diff
@@ -2166,4 +2514,331 @@ mod tests {
...
+
+    // T-WEB-0306: Oversized programs rejected
+    #[tokio::test]
+    async fn program_ingest_rejects_oversized_elf() {
+        let (_store, handler) = make_ingest_handler();
+        let big_elf = vec![0u8; MAX_ELF_UPLOAD_SIZE + 1];
+        let body = make_ingest_body(&big_elf, None, None, None);
+        let err = handler.handle_program_ingest(&body).await.unwrap_err();
+        assert_eq!(err.status_code, 413);
+        assert!(err.message.contains("exceeds limit"));
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply crates/sonde-azure-handler/src/lib.rs:2713]: Fixed -- added program_ingest_exact_size_limit_not_rejected_as_too_large test for the exact boundary case.

File: crates/sonde-azure-handler/src/lib.rs [hunk 10]
```diff
@@ -322,6 +339,284 @@ where
...
+                     expected `resident` or `ephemeral`"
+                )));
+            }
+        };
+
+        let source_filename_raw = body
+            .get("source_filename")
+            .and_then(|v| v.as_str())
+            .map(|s| Some(s.to_string()))
+            .unwrap_or(None);
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply crates/sonde-azure-handler/src/lib.rs]: Fixed -- non-string source_filename now returns 400. Added test program_ingest_rejects_non_string_source_filename.

File: crates/sonde-azure-handler/src/lib.rs [hunk 11]
```diff
@@ -322,6 +339,284 @@ where
...
+            source_filename: record.source_filename.clone(),
+            abi_version: record.abi_version,
+            size_bytes: record.size,
+            verification_profile: profile_name.to_string(),
+            created_at: now,
+        };
+        self.store
+            .store_program_image(&row)
+            .await
+            .map_err(|e| IngestError::internal(format!("store program failed: {e}")))?;
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply crates/sonde-azure-handler/src/lib.rs:467]: Fixed -- added program_ingest_store_failure_returns_500 test using a FailingProgramStore that returns Err from store_program_image.

File: docs/web-ui-validation.md [hunk 12]
```diff
@@ -22,14 +22,14 @@
...
-| T-WEB-0306 | WEB-0306 | Oversized programs rejected | Unit (Rust) | Planned |
-| T-WEB-0307 | WEB-0307 | Empty ELF and multi-program ELF rejected | Unit (Rust) | Planned |
-| T-WEB-0308 | WEB-0308 | `source_filename` normalized to basename | Unit (Rust) | Planned |
+| T-WEB-0301 | WEB-0301 | `ProgramIngest` accepts ELF + metadata via JSON POST | Unit (Rust) | Pass |
+| T-WEB-0302 | WEB-0302 | Prevail verification runs; invalid ELF rejected | Unit (Rust) | Pass |
+| T-WEB-0303 | WEB-0303 | Program hash matches gateway computation | Unit (Rust) | Pass |
+| T-WEB-0304 | WEB-0304 | Program stored in `programs` table with all fields | Unit (Rust) | Pass |
+| T-WEB-0305 | WEB-0305 | Success returns hash+metadata; failure returns diagnostics | Unit (Rust) | Pass |
+| T-WEB-0306 | WEB-0306 | Oversized programs rejected | Unit (Rust) | Pass |
+| T-WEB-0307 | WEB-0307 | Empty ELF and multi-program ELF rejected | Unit (Rust) | Pass |
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/web-ui-validation.md]: Fixed -- split T-WEB-0307 into 0307a (empty ELF, Pass) and 0307b (multi-program ELF, Planned).

File: .vscode/settings.json [hunk 13]
```diff
@@ -0,0 +1,3 @@
+{
+    "sarif-viewer.connectToGithubCodeScanning": "off"
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply .vscode/settings.json]: Fixed -- removed .vscode/settings.json from the PR.

File: docs/web-ui-validation.md [hunk 14]
```diff
@@ -22,14 +22,14 @@
...
-| T-WEB-0303 | WEB-0303 | Program hash matches gateway computation | Unit (Rust) | Planned |
-| T-WEB-0304 | WEB-0304 | Program stored in `programs` table with all fields | Integration | Planned |
-| T-WEB-0305 | WEB-0305 | Success returns hash+metadata; failure returns diagnostics | Unit (Rust) | Planned |
-| T-WEB-0306 | WEB-0306 | Oversized programs rejected | Unit (Rust) | Planned |
-| T-WEB-0307 | WEB-0307 | Empty ELF and multi-program ELF rejected | Unit (Rust) | Planned |
-| T-WEB-0308 | WEB-0308 | `source_filename` normalized to basename | Unit (Rust) | Planned |
+| T-WEB-0301 | WEB-0301 | `ProgramIngest` accepts ELF + metadata via JSON POST | Unit (Rust) | Pass |
+| T-WEB-0302 | WEB-0302 | Prevail verification runs; invalid ELF rejected | Unit (Rust) | Pass |
+| T-WEB-0303 | WEB-0303 | Program hash matches gateway computation | Unit (Rust) | Pass |
+| T-WEB-0304 | WEB-0304 | Program stored in `programs` table with all fields | Unit (Rust) | Pass |
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/web-ui-validation.md]: Fixed -- restored T-WEB-0304 to Integration/Planned. The unit test covers the in-memory row but does not exercise AzureTablesStore serialization.

File: crates/sonde-azure-handler/src/lib.rs [hunk 15]
```diff
@@ -322,6 +339,284 @@ where
...
+        let abi_version = match body.get("abi_version") {
+            Some(serde_json::Value::Null) | None => None,
+            Some(v) => {
+                let raw = v.as_u64().ok_or_else(|| {
+                    IngestError::bad_request("`abi_version` must be a non-negative integer")
+                })?;
+                let val = u32::try_from(raw).map_err(|_| {
+                    IngestError::bad_request(format!(
+                        "`abi_version` value {raw} exceeds maximum ({})",
+                        u32::MAX
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply crates/sonde-azure-handler/src/lib.rs]: Fixed -- abi_version now capped at i32::MAX to match Edm.Int32. Added test program_ingest_rejects_abi_version_above_i32_max.

File: docs/gateway-companion-api.md [hunk 16]
```diff
@@ -184,6 +184,7 @@ unchanged".
 | `assigned_program_hash` | 1 | bstr/null | Desired resident program hash. `null` means no resident program assignment is desired. |
 | `schedule_interval_s` | 2 | uint/null | Desired node wake interval in seconds. `null` means no scheduled interval target is desired in this draft. |
 | `ephemeral_program_hash` | 3 | bstr/null | Desired ephemeral program hash to queue when reconciliation determines one is needed. `null` means no ephemeral run is requested. |
+| `assigned_program_image` | 5 | bstr/null | CBOR program image bytes for the assigned program. Present when the control plane embeds the program image inline so the gateway can ingest it into its local `ProgramLibrary` without a separate fetch. Absent or `null` means no inline image is provided; the gateway must already have the program locally. |
 
 Fields not yet defined by this draft remain reserved for future desired-state
 extension. Receivers MUST ignore unknown integer keys.
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/gateway-companion-api.md:190]: Fixed -- updated the note to say key 5 is now documented in gateway-companion-api.md; only the connector.rs parsing remains as follow-up.

File: crates/sonde-azure-handler/src/lib.rs [hunk 17]
```diff
@@ -2166,4 +2522,439 @@ mod tests {
...
+        }
+        if let Some(profile) = verification_profile {
+            body["verification_profile"] = serde_json::json!(profile);
+        }
+        body
+    }
+
+    // T-WEB-0301: ProgramIngest accepts ELF + metadata via JSON POST
+    // T-WEB-0304: Program stored in programs table with all fields
+    #[tokio::test]
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply crates/sonde-azure-handler/src/lib.rs:2628]: Fixed -- removed T-WEB-0304 label; comment now says partial coverage toward T-WEB-0304 with full integration coverage planned.

File: crates/sonde-azure-handler/src/lib.rs [hunk 18]
```diff
@@ -322,6 +339,289 @@ where
...
+    }
+}
+
+impl std::fmt::Display for IngestError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "HTTP {}: {}", self.status_code, self.message)
+    }
+}
+
+fn chrono_iso8601_utc_now() -> String {
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply crates/sonde-azure-handler/src/lib.rs:532]: Fixed -- added chrono_iso8601_utc_now_has_valid_format test that validates ISO 8601 shape (length, separators, Z suffix, year range).

File: crates/sonde-azure-handler/src/lib.rs [hunk 19]
```diff
@@ -322,6 +339,289 @@ where
...
+            Some(v) => {
+                let raw = v.as_u64().ok_or_else(|| {
+                    IngestError::bad_request("`abi_version` must be a non-negative integer")
+                })?;
+                // Azure Tables stores this as Edm.Int32, so cap at i32::MAX.
+                let max_abi = i32::MAX as u64;
+                if raw > max_abi {
+                    return Err(IngestError::bad_request(format!(
+                        "`abi_version` value {raw} exceeds Edm.Int32 maximum ({max_abi})"
+                    )));
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply crates/sonde-azure-handler/src/lib.rs:429]: Fixed -- added program_ingest_accepts_abi_version_at_i32_max boundary test.

File: docs/web-ui-validation.md [hunk 20]
```diff
@@ -22,14 +22,15 @@
...
+| T-WEB-0303 | WEB-0303 | Program hash matches gateway computation | Unit (Rust) | Pass |
 | T-WEB-0304 | WEB-0304 | Program stored in `programs` table with all fields | Integration | Planned |
-| T-WEB-0305 | WEB-0305 | Success returns hash+metadata; failure returns diagnostics | Unit (Rust) | Planned |
-| T-WEB-0306 | WEB-0306 | Oversized programs rejected | Unit (Rust) | Planned |
-| T-WEB-0307 | WEB-0307 | Empty ELF and multi-program ELF rejected | Unit (Rust) | Planned |
-| T-WEB-0308 | WEB-0308 | `source_filename` normalized to basename | Unit (Rust) | Planned |
+| T-WEB-0305 | WEB-0305 | Success returns hash+metadata; failure returns diagnostics | Unit (Rust) | Pass |
+| T-WEB-0306 | WEB-0306 | Oversized programs rejected | Unit (Rust) | Pass |
+| T-WEB-0307a | WEB-0307 | Empty ELF rejected | Unit (Rust) | Pass |
+| T-WEB-0307b | WEB-0307 | Multi-program ELF rejected | Unit (Rust) | Planned |
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/web-ui-validation.md:32]: Fixed -- updated PR description to accurately list covered tests and note T-WEB-0304 and T-WEB-0307b as planned.

File: docs/web-ui-design.md [hunk 21]
```diff
@@ -69,15 +69,63 @@ deploy/web-ui/
...
+- Error responses use HTTP status codes: 400 (malformed JSON, missing `elf` field, invalid base64), 413 (ELF exceeds size limit), 422 (Prevail verification failure, invalid ELF), 500 (storage/internal error)
+
+#### 6.1.1 Custom Handler Routing
+
+The Azure Functions custom handler (`main.rs`) routes `/ProgramIngest` to a
+dedicated `handle_program_ingest` handler, separate from the catch-all
+`/{*path}` route used for queue-triggered connector messages. The handler:
+
+1. Extracts the HTTP trigger envelope from the Azure Functions invocation request
+2. Reads `Data.req.Body` (JSON string) and `Data.req.Headers`
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/web-ui-design.md]: Fixed -- removed Data.req.Headers from the doc step; implementation only reads Data.req.Body.

File: docs/web-ui-design.md [hunk 22]
```diff
@@ -69,15 +69,63 @@ deploy/web-ui/
...
+  4. Parse `verification_profile`; default to `Resident` if absent
+  5. Call `ProgramLibrary::ingest_elf(elf_bytes, profile)` — this verifies with Prevail, extracts bytecode + maps, encodes deterministic CBOR, and computes SHA-256
+  6. Normalize `source_filename` via `normalize_display_filename()`
+  7. Store in `programs` Azure Table via `HandlerStore::store_program_image()`
+  8. Return JSON: `{"program_hash": "hex", "size": N, "abi_version": N, "source_filename": "name"}`
+  9. On failure: return JSON error with diagnostics
+- The uploaded ELF is verified and transformed; the persisted and delivered artifact is the deterministic CBOR program image, not the raw ELF
+- `programs` table schema: `PartitionKey="program"`, `RowKey=hex(program_hash)`, `source_filename`, `abi_version` (`Edm.Int32`), `cbor_image` (base64-encoded CBOR program image), `size_bytes` (`Edm.Int32`, CBOR image byte length), `verification_profile`, `created_at` (ISO 8601 UTC string)
+- Idempotent: re-ingesting the same ELF produces the same program hash; metadata fields (`source_filename`, `abi_version`) are overwritten on re-ingest
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/web-ui-design.md]: Fixed -- expanded idempotency note to list all overwritten fields: source_filename, abi_version, verification_profile, created_at (last-writer-wins).

File: .github/docker/Dockerfile.azure-handler-builder [hunk 23]
```diff
@@ -20,12 +20,14 @@ WORKDIR /sonde
...
 
-RUN mkdir -p /out/UpstreamConnector /pkg \
+RUN mkdir -p /out/UpstreamConnector /out/ProgramIngest /pkg \
     && cp target/release/sonde-azure-handler /out/sonde-azure-handler \
     && chmod 0755 /out/sonde-azure-handler \
     && cp crates/sonde-azure-handler/function-app/host.json /out/host.json \
     && cp crates/sonde-azure-handler/function-app/UpstreamConnector/function.json \
        /out/UpstreamConnector/function.json \
+    && cp crates/sonde-azure-handler/function-app/ProgramIngest/function.json \
+       /out/ProgramIngest/function.json \
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply .github/docker/Dockerfile.azure-handler-builder:30]: Fixed -- added ProgramIngest/function.json to both CI smoke test assertions in azure-bootstrap-container.yml.
